src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -222,41 +297,110 @@ namespace detail
                 sess->attachSessionData(e.message());
                 throw;
             }
+
+            return true;
         }
 
         bool nextImpl() override
         {
             if (next_callback)
                 next_callback(count());
 
-            if (use_external_buffer)
+            if (read_range.end && bytes_read == read_range.end.value())
```

Comment:
Will this condition work correctly if we initially have some offset, so number of bytes read is actually less than content length?